heuristic: make wildcard character configurable

marevers opened the pull request:
Closes #1410 

Changes the route transform heuristic mode to replace route/path components (gibberish and numeric ids) with a configurable single character. `*` remains the default. This improves handling of label selectors, as `*` is a regex token and can interfere with regex handling in Prometheus/Grafana. By providing the option of changing the replacement character, this can be avoided if required or wanted.

marevers commented:
I am not sure if in light of this change the wildCard constant should also be changed to `/##`.

https://github.com/grafana/beyla/blob/a380cd04b200deb6d1cd76a0614cbf49d47e99a3/pkg/transform/routes.go#L42

mariomac commented on pkg/internal/transform/route/cluster.go:
```diff
@@ -61,11 +61,11 @@ func ClusterPath(path string) string {
 			nSegments++
 			if skip {
-				p[sPos] = '*'
+				p[sPos] = '#'
```
To avoid introducing a breaking change, maybe we could do this value as configurable. Defaulting to `*`, we could add a `wildcard_char` property to the Routes configuration section.

grcevski replied:
I think this is a good idea and something @marevers actually proposed in the issue. I think we should make it configurable, but for simplicity sake, we should only allow one character, otherwise the code will become more complex. It's OK to introduce breaking changes in the major versions, however we need to be cognisant that if we introduce this change with new default character we'll temporarily increase the cardinality for all customers when they switch.

marevers replied:
Alright, I can see if I can make it configurable (allowing a choice of a single character) while keeping `*` as the default option. That will remove the breaking change while still allowing the choice of something else.

grcevski commented on pkg/internal/transform/route/cluster.go:
```diff
@@ -45,7 +45,7 @@ func InitAutoClassifier() error {
 
 //nolint:cyclop
-func ClusterPath(path string) string {
+func ClusterPath(path string, replacement string) string {
```
Small nit here, can you please change the function second argument to be a char? That way we don't need to dereference the string multiple times.

marevers replied:
I ended up solving this a little differently with my last commit. If I were to use a rune parameter, it would necessitate another type casting to byte as within the `ClusterPath` function, a byte is required. So I thought it better to use a byte parameter instead and move the deferencing of the string to the `ClusterPath` call within `classifyFromPath`. I hope you're okay with that. Tests were adjusted to account for this change.

grcevski commented on pkg/transform/routes.go:
```diff
@@ -49,6 +49,8 @@ type RoutesConfig struct {
 	IgnorePatterns []string   `yaml:"ignored_patterns"`
 	IgnoredEvents  IgnoreMode `yaml:"ignore_mode"`
+	// Character that will be used to replace route segments
+	WildcardChar string `yaml:"wildcard_char,omitempty"`
```
I'm wondering if it's possible to use 'char' here, I've never tried it. If not then string that's validates works for me.

marevers replied:
@grcevski hmm. I will give that a try. I attempted rune before but the yaml unmarshaller does not like that. It seems like it is not possible to unmarshal directly into the rune type.

marevers replied:
I seemingly misunderstood something - I thought there was some 'char' type I did not know about but apparently for Go that is just 'rune'. I confirmed it does not work to unmarshal directly into rune here so we need to keep it to string.

grcevski reviewed (approved):
LGTM! Thanks @marevers again! This looks good to me, I just left one comment around supplying a char to the replace function rather than a string. If we can make it work with char everywhere that would be even better, I just don't know if it's possible.

marevers commented:
@grcevski I've removed 'will' from lines I changed in the docs that should hopefully now pass the doc linter. I also noticed 'will' is used in many locations in the docs, but I refrained from making any changes anywhere else.

mariomac reviewed (approved):
great! Thank you @marevers
